File: Company.cpp
#include "Company.h"
#include "Employee.h"
#include "Developer.h"


void ShpaginCompany::add_employee()
{
	std::shared_ptr<ShpaginEmployee> e  = std::make_shared<ShpaginEmployee>(ShpaginEmployee());
    e->console_input();
    employees.push_back(e);
    changed = true;
}

void ShpaginCompany::add_developer()
{
    std::shared_ptr<ShpaginEmployee> e = std::make_shared<ShpaginDeveloper>(ShpaginDeveloper());
    e->console_input();
    employees.push_back(e);
    changed = true;
}

std::wostream& operator<< (std::wostream& out, const ShpaginCompany& c)
{
	for (auto& e : c.employees) {
        e->console_output();
        std::wcout << L"\n";
	}
	return out;
}

bool ShpaginCompany::empty() const
{
    return employees.size() == 0;
}

bool ShpaginCompany::saved() const
{
    return !changed;
}

bool ShpaginCompany::has_saved_file() const
{
    return file_name != L"";
}

void ShpaginCompany::set_filename(const std::wstring& fn)
{
    file_name = fn;
}

void ShpaginCompany::set_directory(const std::wstring& dir)
{
    directory = dir;
}

void ShpaginCompany::clear()
{
    employees.clear();
    ShpaginEmployee::reset_max_id();
    changed = false;
}

void ShpaginCompany::clear_company()
{
    clear();
    std::wcout << L"Все данные удалены." << std::endl;
}

void ShpaginCompany::save_to_file()
{
    std::ofstream file(directory + file_name + extention, std::ios::binary);
    if (file.is_open()) {

        boost::archive::binary_oarchive oa(file);
        oa& *this;

        std::wcout << file_name << L": Данные " << employees.size() << L" сотрудников(-а) успешно сохранены!" << std::endl;
        changed = false;
    }
    else {
        std::wcout << L"**Нет такого файла в каталоге\n";
        file_name = L"";
    }
    file.close();
}

void ShpaginCompany::show_saves() const {
    const std::filesystem::path dir{directory};
    std::wcout << L"Существующие сохранения: " << std::endl;
    for (auto const& dir_entry : std::filesystem::directory_iterator{ dir }) {
        if (dir_entry.path().extension() == extention) {
            std::wcout << L" > " << dir_entry.path().filename().replace_extension() << std::endl;
        }
    }
    std::cout << std::endl;
}

void ShpaginCompany::load_from_file() {

    std::ifstream file(directory + file_name + extention, std::ios::binary);
    if (file.is_open()) {
        clear();

        boost::archive::binary_iarchive ia(file);
        ia& *this;
        
        std::wcout << file_name << L": Данные " << employees.size() << L" сотрудников(-а) успешно загружены!" << std::endl;
        changed = false;
    }
    else {
        std::wcout << L"**Нет такого файла в каталоге\n";
        file_name = L"";
    }
    file.close();
}

File: CSharp/Company.cs
using System;
using System.Xml.Serialization;

namespace CSharp
{
    internal class ShpaginCompany
    {
        protected List<ShpaginEmployee> employees;

        public bool Changed { get; set; }
        public bool Empty => employees.Count == 0;
        public bool HasSavedFile => FileName != "";
        public bool Saved => !Changed;
        public string FileName { get; set; }
        public string Directory { get; set; }
        protected const string extention = ".xml";

        public ShpaginCompany()
        {
            employees = [];
            Changed = false;
            FileName = "";
            Directory = "";
        }

        public void add_employee()
        {
            ShpaginEmployee e = new();
            e.console_input();
            employees.Add(e);
            Changed = true;
        }

        public void add_developer()
        {
            ShpaginDeveloper d = new();
            d.console_input();
            employees.Add(d);
            Changed = true;
        }

        public void console_output()
        {
            if (employees.Count > 0)
            {
                foreach (var e in employees)
                {
                    e.console_output();
                    Console.WriteLine();
                }
            }
            else
            {
                Console.WriteLine("*Пусто");
            }
        }

        protected void clear()
        {
            employees.Clear();
            ShpaginEmployee.ResetMaxId();
            Changed = false;
        }

        public void clear_company()
        {
            clear();
            Console.WriteLine("Все данные удалены.");
        }

        public void save_to_file()
        {
            var xs = new XmlSerializer(typeof(List<ShpaginEmployee>), new[] { typeof(ShpaginEmployee), typeof(ShpaginDeveloper) });
            using (Stream fs = new FileStream(Directory + FileName + extention, FileMode.OpenOrCreate))
            {
                xs.Serialize(fs, employees);
            }
            Changed = false;
        }

        public bool show_saves()
        {
            var files = System.IO.Directory.GetFiles(Directory, "*" + extention);
            Console.WriteLine("Существующие сохранения:");

            if (files.Length > 0)
            {
                foreach (var file in files)
                {
                    string fileNameWithoutExtension = Path.GetFileNameWithoutExtension(file);
                    Console.WriteLine(" > " + fileNameWithoutExtension);
                }
                Console.WriteLine();
                return true;
            }
            else
            {
                Console.WriteLine("*Пусто");
                return false;
            }
        }

        public void load_from_file()
        {
            var xs = new XmlSerializer(typeof(List<ShpaginEmployee>), new[] { typeof(ShpaginEmployee), typeof(ShpaginDeveloper) });
            using (Stream fs = new FileStream(Directory + FileName + extention, FileMode.Open))
            {
                employees = (xs.Deserialize(fs) as List<ShpaginEmployee>)!;
            }
            Changed = false;
        }
    }
}
